File: 프로그래머스/2/42584. 주식가격/주식가격.cpp
#include <string>
#include <vector>
#include <stack>

using namespace std;

vector<int> solution(vector<int> prices) {
    int n = prices.size();
    vector<int> answer(n, 0);
    stack<int> st;
    
    for (int j = 0; j < n; j++) {
        while (!st.empty() && prices[st.top()] > prices[j]) {
            int i = st.top();
            st.pop();
            answer[i] = j - i;
        }
        st.push(j);
    }
    
    while (!st.empty()) {
        int i = st.top();
        st.pop();
        answer[i] = (n - 1) - i;
    }
    
    return answer;
    
    return answer;
}

File: 프로그래머스/2/42584. 주식가격/주식가격.java
import java.util.*;
import java.util.stream.*;

class Solution {    
    public int[] solution(int[] prices) {
        int n = prices.length;
        ArrayList<Integer> answer = new ArrayList(Collections.nCopies(n, 0));
        Stack<Integer> st = new Stack();
        
        for (int i = 0; i < n; i++) {
            while (!st.empty() && prices[st.peek()] > prices[i]) {
                int index = st.pop();
                answer.set(index, i - index);
            }
            
            st.push(i);
        }
        
        while (!st.empty()) {
            int index = st.pop();
            answer.set(index, (n - 1) - index);
        }
        
        
        return answer.stream().mapToInt(Integer::valueOf).toArray();
    }
}